[ES|QL] Fix autocomplete with incompatible args

dej611 opened the pull request:
## Summary

Fixes #180730

I've improved the autocomplete logic a bit here with the new features from `main`:
* the `constantOnly` check has been pushed down rather than depend on an empty string
* autosuggestion takes into account the previous argument types now to propose the next one



### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios


<!--ONMERGE {"backportTargets":["8.14"]} ONMERGE-->

elasticmachine commented:
Pinging @elastic/kibana-esql (Team:ESQL)

drewdaemon commented:
Awesome—glad to see this. But, you may have to pull in `main` because `auto_bucket` looks quite different now.

dej611 commented:
Synced with `main` and improved the logic there with b70e9bf71a855be9e64402b4e7984bf209be17a8.
Description updated.

drewdaemon reviewed (commented):
The original bug is definitely fixed. Many thanks for that.

I'm not so sure on the new suggestions. On the one hand, I like that they are another indicator of the type of the parameter.

On the other hand, I'm concerned they'll be difficult to interpret.

Take these number suggestions here

<img width="659" alt="Screenshot 2024-04-16 at 12 54 33 PM" src="https://github.com/elastic/kibana/assets/315764/19027de2-3cf3-4e49-89b6-1616cd200f5a">

Does this mean that those are the only acceptable values? That's what the suggestions often mean, including when the user types `mv_sort(field, ` and gets `asc` and `desc`.

Why `10`, `20`, `30`? Are these good choices for `bucket`? What about for other functions in the future?

<img width="621" alt="Screenshot 2024-04-16 at 12 55 01 PM" src="https://github.com/elastic/kibana/assets/315764/0a987878-c7d8-40be-b5f4-14656ce95262">

Or here, where did this date come from? Why is it a good suggestion? Does it matter which of the two I pick?

Given this, I would prefer we scope this PR to the bugfix https://github.com/elastic/kibana/issues/180730 so we can get that in for 8.14.

dej611 commented:
> Does this mean that those are the only acceptable values? That's what the suggestions often mean, including when the user types `mv_sort(field, ` and gets `asc` and `desc`.

We had a similar logic for `limit` where we propose only 3 values. I think user understands that the options are not limited for that. Same applies for date math where only the single unit is proposed, rather than any multiple of them.

`ASC` and `DESC` are an enum case like `chrono_literals`.

> Or here, where did this date come from? Why is it a good suggestion? Does it matter which of the two I pick?

This is implementing a similar logic as in formula absolute time shift. But here I've limited to the current time rather than propose any possible shift. I can see how this is a bit more opinionated, but while testing it I found it convenient as a quick placeholder rather than typing the full ISO date manually.

dej611 commented:
The problem with removing the suggestions is that without them it feels the autocomplete is broken as nothing will be then suggested for those argument, as with the fix the `constantsOnly` is enforced in the suggestion system.

drewdaemon commented:
> The problem with removing the suggestions is that without them it feels the autocomplete is broken as nothing will be then suggested for those argument, as with the fix the constantsOnly is enforced in the suggestion system.

I see what you mean. We suggest for pretty much everything else, so this is more jarring. But is it the end of the world? The user still has the inline documentation. The validation is still correct.

I'm not saying that constant suggestions are a bad idea. I'm just saying, why not take them as an enhancement request separate from this bugfix that I'd like to see in 8.14? That way, we can have time to evaluate the request, and if we decide to add, we can think through the cases and add them with some polish.

I think that there are other solutions to this problem as well. For example, https://github.com/elastic/kibana/issues/180528 would also address this, by giving the user a suggestion of the data type, if not the actual values.

But, cc @stratoula for a second opinion.

drewdaemon commented:
> I've implemented the signature help in the past (another issue https://github.com/elastic/kibana/issues/166089) - it's just 3 lines of code - and it doesn't solve the problem as it's only available for multiline editor (due to https://github.com/elastic/kibana/issues/166085 ).

All decisions can be reversed 🤷

drewdaemon commented:
Synced offline and decided with @stratoula and @dej611  to pursue the literal suggestions but to split them out into their own PR.

drewdaemon reviewed (commented):
Works great in the browser but I think the tests are passing by accident.

<img width="1458" alt="Screenshot 2024-04-18 at 7 48 12 AM" src="https://github.com/elastic/kibana/assets/315764/0afd808f-4026-419f-8a0e-fd28f5d47b37">

You can see the problem by adding a conditional breakpoint (`statement === 'from a | eval st_intersects(field, )'`) at `autocomplete.test.ts:253`. 

The expected suggestions should be scoped down to either cartesian or geo points but instead we expect everything. The reason the test passes is that it generates no suggestions `[]`.

All the assertions are nested in a `for` loop on `suggestions` so it being empty causes no assertions to be run. That makes the test pass.

We should probably
- update the expected suggestions to be correct
- figure out why the `suggest` fn is returning an empty array
- add an array length assertion so that we don't get any more of these quiet failures

Another kind of strange thing (maybe not related to the PR) is that when I add a second comma, I get a list of everything as a suggestion

https://github.com/elastic/kibana/assets/315764/db3c08f5-5fe6-42dd-9384-a9df2bbdcc57

dej611 commented on packages/kbn-esql-validation-autocomplete/src/autocomplete/autocomplete.ts:
```diff
@@ -1142,9 +1164,11 @@ async function getFunctionArgsSuggestions(
         {
-          functions: true,
-          fields: true,
+          // @TODO: improve this to inherit the constant flag from the outer function
+          functions: !shouldBeConstant,
+          fields: !shouldBeConstant,
           variables: variablesExcludingCurrentCommandOnes,
+          literals: shouldBeConstant,
```
No, I think it still makes sense to pass it down.

dej611 commented:
> All the assertions are nested in a `for` loop on `suggestions` so it being empty causes no assertions to be run. That makes the test pass.

Good catch. Will fix it.

dej611 commented:
Addressed all issues raised.
I'm not really a fan of the current experience with `bucket` which leaves the user on its own after the first argument, in particular for the 3rd and 4th arg.

drewdaemon commented:
Sorry, I still don't think the test is working correctly... maybe I'm missing something.

The signatures for `st_intersects` are the following

<img width="500" alt="Screenshot 2024-04-18 at 9 34 02 AM" src="https://github.com/elastic/kibana/assets/315764/97076e1d-86b4-4277-8640-1fb85cd27617">

So, cartesian and geo are never mixed.

If you look at the expected suggestions that are generated for `"from a | eval st_intersects(field, )"`, they include both cartesian and geo, even though the first field has already been populated.

<img width="321" alt="Screenshot 2024-04-18 at 9 30 34 AM" src="https://github.com/elastic/kibana/assets/315764/be3249d8-cf6f-4f1e-b892-c9ff1c92c2f8">

Shouldn't we have just one or the other depending on the type of `field`?

> I'm not really a fan of the current experience with bucket which leaves the user on its own after the first argument, in particular for the 3rd and 4th arg.

Agreed. I thought you were planning on splitting this into a separate PR?

dej611 commented:
> Shouldn't we have just one or the other (geo or cartesian) depending on the type of field?

The bug is in the test, as the passed `field` in the `from a | eval st_intersects(field, )` doesn't hold any type information, so the `existingTypes` is always empty. Adding it here is not something quick, so maybe in a followup is best. I think validation testing has already some code to do it that can be borrowed.

> Agreed. Are you still planning on adding the new suggestions in another PR?

It depends on the time frame and how much it would be quick to be defined and added.

drewdaemon reviewed (approved):
Okay, I will fix the test in another PR. Thanks!

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.14|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/181289">](https://github.com/elastic/kibana/pull/181289)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
